partitura/utils/fluidsynth.py

```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+"""
+This module contains methods for synthesizing score- or performance-like
+objects using fluidsynth. Fluidsynth is an optional dependency.
+"""
+
+import os
+from collections import defaultdict
+from typing import Callable, Optional, Union
+
+import numpy as np
+import partitura as pt
+
+try:
+    from fluidsynth import Synth
+
+    HAS_FLUIDSYNTH = True
+except ImportError:  # pragma: no cover
+    Synth = None  # pragma: no cover
+    HAS_FLUIDSYNTH = False  # pragma: no cover
+
+from partitura.utils.synth import SAMPLE_RATE
+from partitura.performance import PerformanceLike
+from partitura.score import ScoreLike
+from partitura.utils.misc import PathLike, download_file
+from partitura.utils.music import (
+    ensure_notearray,
+    get_time_units_from_note_array,
+    performance_notearray_from_score_notearray,
+)
+
+# MuseScore's soundfont distributed under the MIT License.
+# https://ftp.osuosl.org/pub/musescore/soundfont/MuseScore_General/MuseScore_General_License.md
+DEFAULT_SOUNDFONT_URL = "ftp://ftp.osuosl.org/pub/musescore/soundfont/MuseScore_General/MuseScore_General.sf3"
+
+DEFAULT_SOUNDFONT = os.path.join(
+    pt.__path__[0],
+    "assets",
+    "MuseScore_General.sf3",
+)
+
+if not os.path.exists(DEFAULT_SOUNDFONT) and HAS_FLUIDSYNTH:  # pragma: no cover
+    print(f"Downloading soundfont from {DEFAULT_SOUNDFONT_URL}...")  # pragma: no cover
+    download_file(
+        url=DEFAULT_SOUNDFONT_URL,
+        out=DEFAULT_SOUNDFONT,
+    )  # pragma: no cover
+
+
+def synthesize_fluidsynth(
+    note_info: Union[ScoreLike, PerformanceLike, np.ndarray],
+    samplerate: int = SAMPLE_RATE,
+    soundfont: PathLike = DEFAULT_SOUNDFONT,
+    bpm: Union[float, np.ndarray, Callable] = 60,
+) -> np.ndarray:
+    """
+    Synthesize partitura object with note information using
+    fluidsynth.
+
+    Parameters
+    ----------
+    note_info : ScoreLike, PerformanceLike or np.ndarray
+        A partitura object with note information.
+
+    samplerate: int
+        The sample rate of the audio file in Hz.
+
+    soundfont: PathLike
+        The path to the soundfont (in SF2/SF3 format).
+
+    bpm : float, np.ndarray or callable
+        The bpm to render the output (if the input is a score-like object).
+        See `partitura.utils.music.performance_notearray_from_score_notearray`
+        for more information on this parameter.
+
+    Returns
+    -------
+    output_audio_signal : np.ndarray
+       Audio signal as a 1D array.
+    """
+
+    if not HAS_FLUIDSYNTH:
+        raise ImportError("Fluidsynth is not installed!")  # pragma: no cover
+
+    if isinstance(note_info, pt.performance.Performance):
+        for ppart in note_info:
+            ppart.sustain_pedal_threshold = 127
+
+    if isinstance(note_info, pt.performance.PerformedPart):
+        note_info.sustain_pedal_threshold = 127
```

Comment:
Should the sustain pedal threshold be an argument with a default value of 127? (just a suggestion maybe it is not applicable)

Reply:
In this case, the reason for setting the pedal threshold to 127 is to have correct synthesis of note off and pedal messages. Otherwise, the synthesized signal would use "sound off", and create a very wet version of the performance (since both sustain pedal and sound off would produce a compounded effect).